### ISSUE TITLE
Require both version and overall profile IDs in category?

### ISSUE THREAD

**Issue by fugu13** [adlnet/xapi-profiles#68]
This would allow a single xAPI query for all statements in all versions of a profile. The key issues are:

1) is this worth making it so that all statements so far sent with profiles like cmi5 and the scorm profile no longer follow the rules (and those specs need to be updated)?

2) given most profiles are likely to have few versions, is it an acceptable burden to require tools wanting all statements for any version of the profiles to first look up what versions the profile has then do the separate retrievals?

**Comment by creighton** [comment]
For context: https://github.com/adlnet/xAPI-SCORM-Profile/blob/master/xapi-scorm-profile.md#context-activities . The SCORM Profile required that all statements would include a context activity with an ID of 'https://w3id.org/xapi/scorm'. This allowed consumers to easily recognize SCORM Profile statements, and to easily query the LRS for SCORM Profile statements.

I prefer keeping the SCORM profile base ID 'https://w3id.org/xapi/scorm'. That ID was intended to be the base (overall) profile ID. It's the change to support of versioning profiles via the IRI that is messing it up.  I do sympathize with the dilemma though. Existing SCORM Profile statements would be considered version 1 but there is no way to indicate that short of reissuing corrected statements. I don't have a great answer for that, but my vote is still including the base ID in all statements.

**Comment by fugu13** [comment]
@creighton it would still be possible to have that be the ID used for the SCORM profile in category (at least, for code following the current version/first profile spec version) with using only versioned profiles. The approach that allows that, which is what I'm recommending for most profiles with existing category URIs, is to make that the URI of their first version.

**Comment by jhaag75** [comment]
Let's break this down, at least for my own understanding. I'm trying to wrap my head around versioning of profiles. Let's say that:

1) LRP 1 creates xAPI content that uses SCORM profile (the current version). The context category id (aka base profile id) is https://w3id.org/xapi/scorm

2) LRP 2 creates xAPI content that uses the future version of the SCORM profile v.1.1. The context category id is https://w3id.org/xapi/scorm/v1.1

3) LRP 3 creates xAPI content that uses another future version of the SCORM Profile, v. 1.2. The context category is https://w3id.org/xapi/scorm/v1.2

Which of the following examples are bad or good and why? And is there another option than the two below?

1. **VERSIONING SPECIFIED IN IRI.** When a HTTP GET request is made to the IRI specified in context category, it always returns that version of the profile in HTML or constrained JSON-LD (depending on what type was requested). For example, if https://w3id.org/xapi/scorm/v1.1 were specified then that is what version of the Profile the Statement was based on. In other words, Statement validation against the profile is dependent upon the version specified in the IRI in context category. 

2. **LATEST VERSION ALWAYS RETURNED BY DEFAULT.** When a HTTP GET request is made to the profile id specified in context category, it always returns THE LATEST version of the profile in HTML or constrained JSON-LD. Within the HTML and JSON-LD it specifies (or links to) the versioning changes/history. Let's say the latest version is v.1.2. Well if the constant/base SCORM profile id (https://w3id.org/xapi/scorm) is requested via GET then will return the latest version, v.1.2. However, lets say someone was supporting an older implementation and needed to validate against the first version(v1.0). Inside the JSON-LD of the profile there's an RDF triple that the latest version is a prov:wasRevisionOf the earlier version (v.1.1), which is a prov:wasRevisionOf the first version (v.1.0), https://w3id.org/xapi/v.1.0. In other words the first 1.0 version would have a versioned IRI and could be discovered through the links in the latest version of the profile. The first version would be archived and start the versioning process as soon as a new version is created. Statement validation against a profile will always be against the current version (by default). Validating against older versions would require additional lookup against the profile's versioning history. 

-------------------------
Summary: There are probably many pros and cons. Here are some that I'm thinking of: 

Approach 1: Authors of profiles will have to maintain multiple file versions, and constantly update all of their documentation and examples, etc. above to always use the latest version IRI for the profile.  In terms of maintenance that could be a pain and in terms of testing and validation we could get quite a few people making mistakes if documentation isn't properly maintained. 

Approach 2: Authors of profiles will only need to maintain multiple versions and link to them within the Profile metadata (JSON linked data). A human or machine interpreting the versioning information should be able to obtain older versions by "following their nose" to locate the archived versions. With this approach they wouldn't need to also update all of their documentation as IRIs change. 

I could be wrong on these assumptions. Please correct these or advise. Regardless, versioning is a big responsibility and can negatively impact Profile conformance... so we need to figure this out. I think we need some working examples and tests to help realize the best solution.

**Comment by fugu13** [comment]
Hi Jason,

Step 1 has a misstatement. The context category ID would be https://w3id.org/xapi/scorm, but that's not the base profile ID, that's the profile version ID of the first version. I'm proposing the base ID for the SCORM profile be made something like https://w3id.org/xapi/scorm/base to allow this and make existing statements be profile-following.

Yes, if the IRI in the context category returns a profile, it should return the version of the profile that matches that version. Luckily, the profile returned will have enough metadata that, if it doesn't, the receiving system can check it got the right profile (by checking that the most recent version in the versions metadata is the one desired).

I don't understand what you're saying with Approach 2 and the theoretical cons; no matter what the URI returns, if the most applicable version URI isn't specified in the statement somewhere, it isn't possible to figure out which version to apply, and if that version is there, people updating documentation to reflect new versions will need to happen. That is, approach 2 only works if there's both a base URI and a version URI, entailing just as much work to keep updating docs tracking the most recent version of the profile as Approach 1 but also cluttering up the statement with an extra unnecessary URI (the base URI).

**Comment by fugu13** [comment]
Also, there'd still be an overall IRI for the profile that should host the most recent version (for humans or for machines) of the profile, it just isn't what would go in statements. Documentation that just wanted to indicate where to go to learn about the profile would link to that IRI. Only documentation that showed statement construction would need to be updated (but really only if it were being updated to follow the current version; documentation still about prior versions would leave in the prior version IRI, of course).

**Comment by jhaag75** [comment]
Thanks Russell. So are you proposing something similar to Approach 1? Would this mean there would be two context category properties (one with a base URI and one with a version URI)? Can you paste an example below of what you are proposing?

**Comment by fugu13** [comment]
I'm proposing, as #72 says, there be only one context category property, the version that applies to the statement. That's required no matter what, since we need to know which version applies to the statement in order to apply the rules for that version. A system wanting to validate the statement would need to retrieve the version (either from the URI or a central repository), then apply the rules, and they'd be able to do with the single URI.

Example contextActivities:
```javascript
{"category": [{"id": "https://w3id.org/xapi/scorm"}]}
```

Where that URI is for the first version of the SCORM profile, modeled in a profile that might look (only with all the other properties too) like:

```javascript
{
    "@id": "https://w3id.org/xapi/scorm/base",
    "versions": [
        {
            "@id": "https://w3id.org/xapi/scorm",
            "generatedAtTime": "..."
        }
    ]
}
```

This version would be hosted at the version URL, and the base URL would redirect to the version URL as long as this was the current version, or alternatively also host the version (depending on the model chosen).

The reason the root object has the base URI rather than the versions is because the old version is, in most senses, no longer real -- it has been replaced by the new version. In the present, the base profile and the profile version are the same thing. Only when we need historical information such as for validation do we retrieve the old versions. If the question is just, roughly, "what're the concepts in this profile" we can ask that with the general profile URI instead of the version URI.

**Comment by fugu13** [comment]
This is closed by #72
